File: conformance/tests/covenant-vault/covenant-vault.runar.zig
const runar = @import("runar");

pub const CovenantVault = struct {
    pub const Contract = runar.SmartContract;

    owner: runar.PubKey,
    recipient: runar.Addr,
    minAmount: i64,

    pub fn init(owner: runar.PubKey, recipient: runar.Addr, minAmount: i64) CovenantVault {
        return .{
            .owner = owner,
            .recipient = recipient,
            .minAmount = minAmount,
        };
    }

    pub fn spend(self: *const CovenantVault, sig: runar.Sig, txPreimage: runar.SigHashPreimage) void {
        runar.assert(runar.checkSig(sig, self.owner));
        runar.assert(runar.checkPreimage(txPreimage));

        const p2pkhScript = runar.cat(runar.cat("1976a914", self.recipient), "88ac");
        const expectedOutput = runar.cat(runar.num2bin(self.minAmount, 8), p2pkhScript);

        runar.assert(runar.bytesEq(runar.hash256(expectedOutput), runar.extractOutputHash(txPreimage)));
    }
};

File: conformance/tests/covenant-vault/covenant-vault.runar.go
//go:build ignore

package contract

import "runar"

type CovenantVault struct {
	runar.SmartContract
	Owner     runar.PubKey `runar:"readonly"`
	Recipient runar.Addr   `runar:"readonly"`
	MinAmount runar.Bigint `runar:"readonly"`
}

func (c *CovenantVault) Spend(sig runar.Sig, txPreimage runar.SigHashPreimage) {
	runar.Assert(runar.CheckSig(sig, c.Owner))
	runar.Assert(runar.CheckPreimage(txPreimage))
	p2pkhScript := runar.Cat(runar.Cat("1976a914", c.Recipient), "88ac")
	expectedOutput := runar.Cat(runar.Num2Bin(c.MinAmount, 8), p2pkhScript)
	runar.Assert(runar.Hash256(expectedOutput) == runar.ExtractOutputHash(txPreimage))
}
